### PR TITLE
Hooking backward for QNNPACK

### PR DESCRIPTION
**Pull request by kwanghoon-meta**
Summary: Enabling quantized gradient.

Test Plan:
Algorithmic correctness - Dequantized matmul vs QNNPACK matmul for gradient - P616202766

```
dequantized matmul : [1.5463, -0.2917, -2.1735, 0.5689, -1.0795]
QNNPACK matmul : tensor([[ 1.5463, -0.2917, -2.1735,  0.5689, -1.0795]])
```

Differential Revision: D42593235

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: torch/csrc/autograd/quantized/quantized_backward.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
+#include <ATen/native/quantized/PackedParams.h>
```
Comment (albanD) [inline torch/csrc/autograd/quantized/quantized_backward.cpp:1]: Could you move this to `torch/csrc/quantized/*` to make codeowners maintenance easier?

Comment (kwanghoon-meta) [reply torch/csrc/autograd/quantized/quantized_backward.cpp:1]: alright let me change

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 2]
```diff
@@ -58,6 +58,12 @@
 #define int_vector __m256i
 #endif // CPU_CAPABILITY_AVX512
 
+#if defined(__clang__)
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Werror"
```
Comment (albanD) [inline aten/src/ATen/cpu/vec/vec_base.h:63]: ignoring warnings is not ok. We actually work hard to make this code warning free
cc @malfet

File: buckbuild.bzl [hunk 3]
```diff
@@ -1463,6 +1463,7 @@ def define_buck_targets(
             "torch/csrc/jit/mobile/train/random.cpp",
             "torch/csrc/jit/mobile/train/sequential.cpp",
             ":gen_aten_libtorch[autograd/generated/Functions.cpp]",
+            "torch/csrc/quantized/quantized_backward.cpp",
```
Comment (albanD) [inline buckbuild.bzl:1466]: I'm still surprised this is not in tools/build_variables.bzl. But if you only need it here it's ok. But that means that this won't be included in other targets and so you might get strange behavior when these backwards are not registered there.

File: torch/csrc/quantized/quantized_backward.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            .typed<at::Tensor(
+              at::Tensor,
+              c10::intrusive_ptr<LinearPackedParamsBase,
+              c10::detail::intrusive_target_default_null_type<LinearPackedParamsBase>> const&,
+              bool)>();
+    auto output = op.redispatch(
+      DispatchKeySet({DispatchKey::CPU}), input, packed_weight, reduce_range);
+    auto [weight, bias] = packed_weight->unpack();
+    TORCH_CHECK(weight.qscheme() == c10::kPerTensorAffine);
+    // TO-DO : passing a bias does not work, hence create an arbitrary one in backward loop
```
Comment (albanD) [inline torch/csrc/quantized/quantized_backward.cpp:33]: If you need to save non-Tensor data, you can do so by doing `ctx->saved_data["foo"] = foo`

File: torch/csrc/quantized/quantized_backward.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    c10::optional<at::Tensor> bias;
+    auto prepacked_weight = op.redispatch(
+      DispatchKeySet({DispatchKey::QuantizedCPU}), original_weight, bias);
+    auto grad_input = prepacked_weight->apply_dynamic(grad_output);
+    return {grad_input, torch::Tensor(), torch::Tensor()};
+  }
+};
+
+at::Tensor packed_linear_weight_grad(
+    at::Tensor input,
```
Comment (albanD) [inline torch/csrc/quantized/quantized_backward.cpp:58]: This should take `c10::DispatchKeySet ks, ` as the first argument. And this keyset should be used when redispatching during the forward pass. similar to https://github.com/pytorch/pytorch/blob/7a192cc51c7172860efd35413acf9a8b9aafd2c9/torch/csrc/autograd/VariableTypeManual.cpp#L110-L122

File: torch/csrc/quantized/quantized_backward.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    ctx->save_for_backward({weight});
+    return output;
+  }
+
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs) {
+    auto saved = ctx->get_saved_variables();
+    auto original_weight = saved[0];
+    original_weight = at::permute(original_weight, {1, 0});
+    auto grad_output = grad_outputs[0];
+    static auto op =
```
Comment (albanD) [inline torch/csrc/quantized/quantized_backward.cpp:43]: This is just calling a regular function, you can do it via `quantized::linear_prepack(original_weight, bias);`

The forward above is special because it is called while going down the dispatcher. So it needs to nicely redispatch.
This one is called by the autograd engine as regular code. So you can just do regular calls.

File: aten/src/ATen/cpu/vec/vec_base.h (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/cpu/vec/vec_base.h:63]: Why do you need to touch this file btw?

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 7]
```diff
@@ -7,6 +7,14 @@
 #include <ATen/cpu/vec/vec.h>
 #include <c10/util/irange.h>
 
+// suppress warnings triggered by #pragma unroll:
```
Comment (kimishpatel) [inline aten/src/ATen/native/cpu/CatKernel.cpp:10]: Same for this one. If this has been working before I dont see why this diff has to introduce these changes.

File: aten/src/ATen/native/native_functions.yaml [hunk 8]
```diff
@@ -6347,7 +6347,7 @@
   device_check: NoCheck   # TensorIterator
   variants: method, function
   dispatch:
-    CPU, CUDA: zero_
+    CPU, CUDA, QuantizedCPU: zero_
```
Comment (kimishpatel) [inline aten/src/ATen/native/native_functions.yaml:6350]: @albanD do you know why this might be needed? As I commented on PR Zeroing out a float tensor is not the same as filling quantized tensor with just zeros.

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 9]
```diff
@@ -65,13 +66,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
-        TORCH_WARN(
-        "Original weight is freed, we are converting pre-packed weight to original weight.");
-        uint8_t* kernel = w->unpackWeights(w_zero_points.data(), n_elements);
-        at::Tensor original_tensor = at::from_blob(kernel, weight_sizes, c10::kByte).clone().toType(c10::kQInt8);
-        original_tensor.sub_(128);
-        free(kernel);
-        return std::tuple<at::Tensor, c10::optional<at::Tensor>>(original_tensor, bias_);
+        float* weight_scales_data = w_scales.data_ptr<float>();
+        at::Tensor weight_origin;
+        weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:71]: Please use this API instead https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear.cpp#L240.

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 10]
```diff
@@ -65,13 +66,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
+        int8_t* weight_ptr_int8 =
+            reinterpret_cast<int8_t*>(weight_origin.data_ptr<int8_t>());
+        w->unpackWeights(w_zero_points.data(), weight_ptr_int8);
+        // See for the subtraction 128
+        // https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L319
+        weight_origin.sub_(128);
+        at::Tensor original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:78]: Also this is covering strictly per tensor path only, what about per channel. Either assert that the tensor being unpacked is per tensor quantized only or add path for per channel as well.

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 11]
```diff
@@ -65,13 +66,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
+        TORCH_CHECK(original_quantized_tensor.q_scale() == weight_scales_data[0]);
+        TORCH_CHECK(original_quantized_tensor.q_zero_point() == w_zero_points[0]);
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:80]: Why this check? Shouldnt be needed.

File: aten/src/ATen/cpu/vec/vec_base.h (hunk 2 above)
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec_base.h:63]: Please use `C10_CLANG_DIAGNOSTIC_PUSH` and `C10_CLANG_DIAGNOSTIC_IGNORE`, but agree with @albanD that ignoring warnings require some explanation

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 12]
```diff
@@ -7,6 +7,14 @@
...
+// "warning: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]"
+#if defined(__clang__)
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Werror"
+#  pragma clang diagnostic ignored "-Wpass-failed"
+#endif
```
Comment (malfet) [inline aten/src/ATen/native/cpu/CatKernel.cpp:16]: Use convenience macros and never ignore `-Werror` 
```suggestion
C10_CLANG_DIAGNOSTIC_PUSH()
C10_CLANG_DIAGNOSTIC_IGNORE("-Wpass-failed")
```

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not push `-Werror` to ignored settings, and if you have to push some exclusions use `C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED`, which is backward compatible with older clangs where exclusion might not be defined

File: buckbuild.bzl (hunk 3 above)
Comment (kwanghoon-meta) [reply buckbuild.bzl:1466]: Hmmmm where should I place the code then?

File: torch/csrc/quantized/quantized_backward.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch/csrc/quantized/quantized_backward.cpp:33]: @albanD bias is also a tensor.

@kwanghoon-meta note that you are unpacking weights at forward, so how you will have two copies of weights. One in the packed_weight another for gradient that you are saving. So this will increase your memory footprint if that is what you wanted to minimize. Really what you want is to save packed_weight but I dont know if you can do that with autograd. But if this memory increase is fine then this works.

File: torch/csrc/quantized/quantized_backward.cpp (hunk 6 above)
Comment (kimishpatel) [reply torch/csrc/quantized/quantized_backward.cpp:43]: > The forward above is special because it is called while going down the dispatcher. So it needs to nicely redispatch.

@albanD what is the "nicely redispatch" part?

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp (hunk 10 above)
Comment (kwanghoon-meta) [reply aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:78]: This is very first step, we will expand to per channel accordingly. per channel requires more operation implementation in order to go through autograd engine (striding needs to be implemented etc)
Let me add a flag as of now with assert

File: aten/src/ATen/cpu/vec/vec_base.h (hunk 2 above)
Comment (kwanghoon-meta) [reply aten/src/ATen/cpu/vec/vec_base.h:63]: Seems like my old code was creating a bug, which requires such change, now it does not.
Let me remove, thanks for a feedback.

File: buckbuild.bzl (hunk 3 above)
Comment (albanD) [reply buckbuild.bzl:1466]: I'm not 100% sure tbh but I would have put it with similar files in https://github.com/pytorch/pytorch/blob/5b2ab0dd4f9b75657ba0e83641cd91864b4f23d1/build_variables.bzl#L156

File: aten/src/ATen/native/native_functions.yaml (hunk 8 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:6350]: I don't know why this is needed no. Since there is no testing in this PR, I doubt any new codepath is actually exercised so I suspect this is not needed.
If you have specific test that requires this, I'm sure we can figure out why it is needed if it is indeed.

File: torch/csrc/quantized/quantized_backward.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch/csrc/quantized/quantized_backward.cpp:33]: I think the reason passing bias dose not work is because gradient uses transposed weights where as bias shape is aligned for non-tranposed weight. But thinking more why do you need to pass bias at all or even use bias in the backward since gradient of bias addition w.r.t input is really zero, no?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left some comments

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp (hunk 10 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:78]: Then at the least assert that the weight tensor is per channel.

File: torch/csrc/quantized/quantized_backward.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch/csrc/quantized/quantized_backward.cpp:33]: > > Really what you want is to save packed_weight but I dont know if you can do that with autograd.
> 
> Just use `saved_data` ;)

oh thats right. You just said that above.

File: torch/csrc/quantized/quantized_backward.cpp (hunk 6 above)
Comment (albanD) [reply torch/csrc/quantized/quantized_backward.cpp:43]: > @albanD what is the "nicely redispatch" part?

Getting the `DispatchKeySet` argument from the caller and calling `redispatch` with the properly updated keyset.

Since you're not using codegen here, you indeed don't have a `quantized::` namespace. Then you can just update the line below to use `call()` instead of `redispatch()`.

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 13]
```diff
@@ -65,13 +66,20 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
-        TORCH_WARN(
-        "Original weight is freed, we are converting pre-packed weight to original weight.");
-        uint8_t* kernel = w->unpackWeights(w_zero_points.data(), n_elements);
-        at::Tensor original_tensor = at::from_blob(kernel, weight_sizes, c10::kByte).clone().toType(c10::kQInt8);
-        original_tensor.sub_(128);
-        free(kernel);
-        return std::tuple<at::Tensor, c10::optional<at::Tensor>>(original_tensor, bias_);
+        float* weight_scales_data = w_scales.data_ptr<float>();
+        at::Tensor weight_origin;
+        weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
```
Comment (malfet) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:71]:
```suggestion
        auto weight_origin = at::empty(weight_sizes, at::device(c10::kCPU).dtype(at::kChar));
```

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 14]
```diff
@@ -65,13 +66,20 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
+        int8_t* weight_ptr_int8 =
+            reinterpret_cast<int8_t*>(weight_origin.data_ptr<int8_t>());
```
Comment (malfet) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:73]: Why reinterpret cast is needed here?
```suggestion
        int8_t* weight_ptr_int8 = weight_origin.data_ptr<int8_t>();
```

File: aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp [hunk 15]
```diff
@@ -65,13 +66,20 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
...
+        w->unpackWeights(w_zero_points.data(), weight_ptr_int8);
+        // See for the subtraction 128
+        // https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L319
+        weight_origin.sub_(128);
+        // As of now, we are supporting only per tensor quantizer
+        // TO-DO : Support a per channel as well.
+        at::Tensor original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
```
Comment (malfet) [inline aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:80]: Nit
```suggestion
        auto original_quantized_tensor = at::_make_per_tensor_quantized_tensor(weight_origin, weight_scales_data[0], w_zero_points[0]);
```

File: torch/csrc/quantized/quantized_backward.cpp [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+// back to the previous layers
+// This function can be used when the user is adapting mixed precision for traninig after quantization
+// From torch layer, we have no access to linear_dynamic operator which needs to access via redispatching mechanism
+// TO-DO : currently we are supporting per tensor quantization only, will expand to per channel later on
+class PackedLinearWeightDynamicBackward :
+public Function<PackedLinearWeightDynamicBackward> {
+ public:
+  static torch::Tensor forward(
+    AutogradContext *ctx,
+    at::Tensor input,
```
Comment (malfet) [inline torch/csrc/quantized/quantized_backward.cpp:18]: What's the difference between `at::Tensor` and `torch::Tensor`? And why use `at::` namespace if you already have `using namespace at;` on line 7
```suggestion
  static Tensor forward(
    AutogradContext *ctx,
    Tensor input,
```

**Review by malfet — APPROVED** [review]
LGTM, but have a look at few of the nits

**Review by kimishpatel — APPROVED** [review]
